library/src/test/java/com/vinted/coper/CoperImplTest.kt

```diff
@@ -374,25 +375,73 @@ class CoperImplTest {
     @Test
     fun getFragment_twoRequest_sameInstance() {
         runBlocking {
-            val fixture = CoperImpl(activity.supportFragmentManager)
-            val firstFragment = fixture.getFragment()
-            val secondFragment = fixture.getFragment()
+            val fixture = getCoperInstance()
+            val firstFragment = fixture.getFragmentSafely()
+            val secondFragment = fixture.getFragmentSafely()
 
             assertTrue { firstFragment === secondFragment }
         }
     }
 
-    @Test(expected = IllegalStateException::class)
-    fun request_onIoThread_shouldCrash() {
-        val fixture = spy(CoperImpl(activity.supportFragmentManager))
-        fixture.mockGetFragmentWithStub()
+    @Test
+    fun getFragment_manyRequestAsync_onlyOneFragmentCreated() {
+        runBlocking {
+            val activityController = Robolectric.buildActivity(FragmentActivity::class.java)
+            val activity: FragmentActivity = activityController.setup().get()
+            val fragmentManager = activity.supportFragmentManager
+            val fixture = getCoperInstance(
+                lifecycle = activity.lifecycle,
+                fragmentManager = fragmentManager
+            )
+            val fragmentPromises = mutableListOf<Deferred<CoperFragment?>>()
 
-        runBlocking(Dispatchers.IO) {
-            executePermissionRequest(
-                coperImplReference = fixture,
-                permissions = listOf("test"),
-                permissionResult = listOf(PermissionChecker.PERMISSION_GRANTED)
+            repeat(10) {
+                fragmentPromises.add(async(Dispatchers.IO) {
+                    runCatching {
+                        fixture.getFragmentSafely()
+                    }.getOrNull()
+                })
+            }
+
+            fixture.latestCommittedFragmentFlow.filterNotNull().first()
```

Comment:
Here you are checking the latest fragment which was committed. Do you have any test which checks if a fragment was actually added in the fragment manager and used correctly?

Reply:
good catch, I already expected that it works, it would be nice to test it

Reply:
Technically you could use this information, but some tests check if it works correctly with fragment manager as well. But the best way is to write a test is not to depend on implementation details, but rather do some action -> mangle fragment/activity state -> do some action and check the result. Robolectric offers some capabilities of controlling activity state and even restoring from the state. 

Checking implementation details is useful for checking memory leaks and sometimes it is not avoidable for race-condition tests.

Reply:
Also, sometimes it is not avoidable to test things developed on another framework.

Reply:
Now I checked this, what do you mean by:
```
Do you have any test which checks if a fragment was actually added in the fragment manager and used correctly
```

There is tests that check if fragment is added to fragment manager and and it returns fragment that is already added. About what type of tests you are talking here?

Reply:
> About what type of tests you are talking here?

I am talking about tests that do not depend on implementation details. It means, your tests should depend on FragmentManager. The problem with your tests is, even if a fragment was added it does not mean the whole solution is working. Some of your tests even do not check if the fragment was added or used, just depends on how correctly implementation reports about the fragment. 

> There is tests that check if fragment is added to fragment manager and and it returns fragment that is already added.

Then all is good.